WeCantSpell.Hunspell/CharacterCondition.cs

```diff
@@ -10,43 +11,67 @@ namespace WeCantSpell.Hunspell;
 
 public readonly struct CharacterCondition : IReadOnlyList<char>, IEquatable<CharacterCondition>
 {
-    public static readonly CharacterCondition AllowAny = new([], ModeKind.RestrictChars);
+    public static readonly CharacterCondition AllowAny = new(string.Empty, ModeKind.RestrictChars);
 
     public static bool operator ==(CharacterCondition left, CharacterCondition right) => left.Equals(right);
 
-    public static bool operator !=(CharacterCondition left, CharacterCondition right) => !(left == right);
+    public static bool operator !=(CharacterCondition left, CharacterCondition right) => !left.Equals(right);
 
     public static CharacterCondition CreateCharSet(ReadOnlySpan<char> chars, bool restricted)
     {
-        var builder = ArrayBuilder<char>.Pool.Get(chars.Length);
-
-        foreach (var c in chars)
-        {
-            builder.AddAsSortedSet(c);
-        }
+        var builder = new StringBuilderSpan(chars);
+        builder.Sort();
+        builder.RemoveAdjacentDuplicates();
+        return new(builder.GetStringAndDispose(), restricted ? ModeKind.RestrictChars : ModeKind.PermitChars);
+    }
 
-        return new(ArrayBuilder<char>.Pool.ExtractAndReturn(builder), restricted ? ModeKind.RestrictChars : ModeKind.PermitChars);
+    public static CharacterCondition CreateCharSet(string chars, bool restricted)
+    {
+        var charsSpan = chars.AsSpan();
+        return charsSpan.CheckSortedWithoutDuplicates()
+            ? new(chars, restricted ? ModeKind.RestrictChars : ModeKind.PermitChars)
+            : CreateCharSet(charsSpan, restricted);
     }
 
-    public static CharacterCondition CreateSequence(char c) => new([c], ModeKind.MatchSequence);
+    public static CharacterCondition CreateSequence(char c) => CreateSequence(c.ToString());
+
+    public static CharacterCondition CreateSequence(ReadOnlySpan<char> chars) => CreateSequence(chars.ToString());
 
-    public static CharacterCondition CreateSequence(ReadOnlySpan<char> chars) => new(chars.ToArray(), ModeKind.MatchSequence);
+    public static CharacterCondition CreateSequence(string chars) => new(chars, ModeKind.MatchSequence);
 
-    private CharacterCondition(char[] characters, ModeKind mode)
+#if HAS_SEARCHVALUES
+
+    private CharacterCondition(string characters, ModeKind mode)
     {
         _characters = characters;
         _mode = mode;
+        _searchValues = _mode is not ModeKind.MatchSequence
```

Comment:
Search values may not be worth the trouble or branching here. Do some experimentation to be sure.

Reply:
It wasn't helping.